### DISCUSSION TITLE
Failed to open dyn channle in server side.

### DISCUSSION THREAD

**Discussion by RyanCode** [FreeRDP/FreeRDP#9379] [Q&A] [answered]
## What happened？

**I want to build a customed channle between freerdp server and mstsc named rdpgw.I try to call `WTSVirtualChannelOpenEx` In channels_on_client_channel_connect.but return error 203**

## user mstsc
I have build 'rdpgw.dll' and also configured the dynamic library for the channel according to the document https://learn.microsoft.com/en-us/windows/win32/termserv/virtual-channel-client-registration

<img width="714" alt="image" src="https://github.com/FreeRDP/FreeRDP/assets/61410262/38ef83c3-0e05-49b6-8259-aa109609d61f">

```c
	void* rdpgw_channle =
	    WTSVirtualChannelOpenEx(WTS_CURRENT_SESSION, "rdpgw", WTS_CHANNEL_OPTION_DYNAMIC);
	if (!rdpgw_channle)
	{
		WLog_ERR(TAG, "failed to open rdpgw channel:%d", GetLastError());
		return FALSE;
	}
```
plz give me some advice,thanks a lot.

**Comment by hardening** [comment, marked as answer]
TLDR: I'd recommand to directly use the wts lib under windows not FreeRDP's WTSAPI wrapper. There's some known buggy behaviors, especially with dynamic channels.

Long story: in WinPR under win32 we try to bring our enhancements of the API under windows (like having an `Event` that you can `WaitForSingleObject` on to poll channel activity), so we do a complicated mix of calling the original WTSAPI functions with overlaps and completions. It kinda work with static channels, but it's clearly buggy with dynamic ones. I've seen the bug, but don't have enough interest in win32 to find and fix it.

Unfortunately I'm not sure you can have freerdp and windows wtsapi at the same time because FreeRDP headers includes `winpr/wtsapi.h`
